### PR TITLE
Rustdoc-Json: Retain Stripped Modules when they are imported, not when they have items

### LINKED ISSUE #100973: Rustdoc-Json: Reference to missing item with 2 glob imports
```rust
#![feature(no_core)]
#![no_core]
mod m1 {
    pub fn f() {}
}
mod m2 {
    pub fn f(_: u8) {}
}
pub use m2::*;
pub use m1::*;
```

Produces an output with references to non existant items: `0:3:1564` and `0:1:1563`

```json
{
  "crate_version": null,
  "external_crates": {},
  "format_version": 18,
  "includes_private": false,
  "index": {
    "0:0:1565": {
      "crate_id": 0,
      "id": "0:0:1565",
      "inner": {"is_crate": true, "is_stripped": false, "items": ["0:5", "0:6"]},
      "kind": "module",
      "name": "glob_private_simple",
      "visibility": "public"
    },
    "0:5": {
      "crate_id": 0,
      "id": "0:5",
      "inner": {"glob": true, "id": "0:3:1564", "name": "m2", "source": "m2"},
      "kind": "import",
      "name": null,
      "visibility": "public"
    },
    "0:6": {
      "crate_id": 0,
      "id": "0:6",
      "inner": {"glob": true, "id": "0:1:1563", "name": "m1", "source": "m1"},
      "kind": "import",
      "name": null,
      "visibility": "public"
    }
  },
  "root": "0:0:1565"
}
```

<details>

<summary> Full </summary>

```json
{
  "crate_version": null,
  "external_crates": {},
  "format_version": 18,
  "includes_private": false,
  "index": {
    "0:0:1565": {
      "attrs": ["#![feature(no_core)]", "#![no_core]"],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:0:1565",
      "inner": {"is_crate": true, "is_stripped": false, "items": ["0:5", "0:6"]},
      "kind": "module",
      "links": {},
      "name": "glob_private_simple",
      "span": {
        "begin": [1, 0],
        "end": [12, 14],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    },
    "0:5": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:5",
      "inner": {"glob": true, "id": "0:3:1564", "name": "m2", "source": "m2"},
      "kind": "import",
      "links": {},
      "name": null,
      "span": {
        "begin": [11, 0],
        "end": [11, 14],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    },
    "0:6": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:6",
      "inner": {"glob": true, "id": "0:1:1563", "name": "m1", "source": "m1"},
      "kind": "import",
      "links": {},
      "name": null,
      "span": {
        "begin": [12, 0],
        "end": [12, 14],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    }
  },
  "paths": {"0:0:1565": {"crate_id": 0, "kind": "module", "path": ["glob_private_simple"]}},
  "root": "0:0:1565"
}
```

</details>

If you remove one of the imports, everything is fine

```rust
#![feature(no_core)]
#![no_core]
mod m1 {
    pub fn f() {}
}
mod m2 {
    pub fn f(_: u8) {}
}
pub use m2::*;
```

```json
{
  "crate_version": null,
  "external_crates": {},
  "format_version": 18,
  "includes_private": false,
  "index": {
    "0:0:1565": {
      "crate_id": 0,
      "id": "0:0:1565",
      "inner": {"is_crate": true, "is_stripped": false, "items": ["0:5"]},
      "kind": "module",
      "name": "glob_private_simple",
      "visibility": "public"
    },
    "0:3:1564": {
      "crate_id": 0,
      "id": "0:3:1564",
      "inner": {"is_crate": false, "is_stripped": true, "items": ["0:4:630"]},
      "kind": "module",
      "name": "m2",
      "visibility": "crate"
    },
    "0:4:630": {
      "crate_id": 0,
      "id": "0:4:630",
      "inner": { ... },
      "kind": "function",
      "name": "f",
      "visibility": "public"
    },
    "0:5": {
      "crate_id": 0,
      "id": "0:5",
      "inner": {"glob": true, "id": "0:3:1564", "name": "m2", "source": "m2"},
      "kind": "import",
      "name": null,
      "visibility": "public"
    }
  },
  "root": "0:0:1565"
}
```

<details>

<summary> Full </summary>

```json
{
  "crate_version": null,
  "external_crates": {},
  "format_version": 18,
  "includes_private": false,
  "index": {
    "0:0:1565": {
      "attrs": ["#![feature(no_core)]", "#![no_core]"],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:0:1565",
      "inner": {"is_crate": true, "is_stripped": false, "items": ["0:5"]},
      "kind": "module",
      "links": {},
      "name": "glob_private_simple",
      "span": {
        "begin": [1, 0],
        "end": [11, 14],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    },
    "0:3:1564": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:3:1564",
      "inner": {"is_crate": false, "is_stripped": true, "items": ["0:4:630"]},
      "kind": "module",
      "links": {},
      "name": "m2",
      "span": {
        "begin": [7, 0],
        "end": [7, 6],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "crate"
    },
    "0:4:630": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:4:630",
      "inner": {
        "decl": {
          "c_variadic": false,
          "inputs": [["_", {"inner": "u8", "kind": "primitive"}]],
          "output": null
        },
        "generics": {"params": [], "where_predicates": []},
        "header": {"abi": "Rust", "async": false, "const": false, "unsafe": false}
      },
      "kind": "function",
      "links": {},
      "name": "f",
      "span": {
        "begin": [8, 4],
        "end": [8, 22],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    },
    "0:5": {
      "attrs": [],
      "crate_id": 0,
      "deprecation": null,
      "docs": null,
      "id": "0:5",
      "inner": {"glob": true, "id": "0:3:1564", "name": "m2", "source": "m2"},
      "kind": "import",
      "links": {},
      "name": null,
      "span": {
        "begin": [11, 0],
        "end": [11, 14],
        "filename": "/home/nixon/dev/rust/rust/src/test/rustdoc-json/reexport/glob_private_simple.rs"
      },
      "visibility": "public"
    }
  },
  "paths": {"0:0:1565": {"crate_id": 0, "kind": "module", "path": ["glob_private_simple"]}},
  "root": "0:0:1565"
}
```

</details>

This should have been caught by `check_missing_items.py`, but wasn't. Instead is was caught by [`jsondoclint`](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/check_missing_items.2Epy.20Replacement.2E), a rust based replacement (PR soon).

@rustbot modify labels: +A-rustdoc-json +T-rustdoc

### LINKED ISSUE #101103: Rustdoc-Json: Reexporting item in module adds that module to index
```rust
#![feature(no_core)]
#![no_core]

mod m1 {
    pub fn x(){}
}

pub use m1::x;
```

Produces (abbridged)

```json
{
  "crate_version": null,
  "external_crates": {},
  "format_version": 18,
  "includes_private": false,
  "index": {
    "0:0:1568": {
      "crate_id": 0,
      "id": "0:0:1568",
      "inner": {"is_crate": true, "is_stripped": false, "items": ["0:3"]},
      "kind": "module",
      "name": "bug"
    },
    "0:1:1566": {
      "crate_id": 0,
      "id": "0:1:1566",
      "inner": {"is_crate": false, "is_stripped": true, "items": ["0:2:1567"]},
      "kind": "module",
      "name": "m1"
    },
    "0:2:1567": {
      "crate_id": 0,
      "id": "0:2:1567",
      "inner": {"decl": {"c_variadic": false, "inputs": [], "output": null}},
      "kind": "function",
      "name": "x"
    },
    "0:3": {
      "crate_id": 0,
      "id": "0:3",
      "inner": {"glob": false, "id": "0:2:1567", "name": "x", "source": "m1::x"},
      "kind": "import",
      "name": null
    }
  },
  "root": "0:0:1568"
}
```

But theirs no need for `0:1:1566` to be in the index: `0:1:1566`

@rustbot modify labels: +A-rustdoc-json +T-rustdoc

@rustbot claim

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
Fixes #101103
Fixes #100973

r? @GuillaumeGomez

### CODE REVIEW

File: src/librustdoc/json/import_finder.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def_id::DefId;
+
+use crate::{
+    clean::{self, Import, ImportSource, Item},
+    fold::DocFolder,
+};
+
+pub(crate) fn get_imports(krate: clean::Crate) -> (clean::Crate, FxHashSet<DefId>) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/json/import_finder.rs:9]: Please add a small comment explaining what this function does.

**Comment by GuillaumeGomez** [comment]
Apart a little missing documentation (very useful for anyone who comes reading this part of the code), looks all good to me!

File: src/librustdoc/json/import_finder.rs (hunk 1 above)
Comment (aDotInTheVoid) [reply src/librustdoc/json/import_finder.rs:17]: Done

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r- failed in a rollup
https://github.com/rust-lang/rust/pull/101136#issuecomment-1229567751

**Comment by aDotInTheVoid** [comment]
[error](https://github.com/rust-lang-ci/rust/runs/8060819316?check_suite_focus=true)
```
 Documenting rustdoc v0.0.0 (/checkout/src/librustdoc)
error: this URL is not a hyperlink
  --> src/librustdoc/json/import_finder.rs:14:9
   |
14 | /// See https://github.com/rust-lang/rust/issues/100973 and
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use an automatic link instead: `<https://github.com/rust-lang/rust/issues/100973>`
   |
   = note: `-D rustdoc::bare-urls` implied by `-D warnings`
   = note: bare URLs are not automatically turned into clickable links

error: this URL is not a hyperlink
  --> src/librustdoc/json/import_finder.rs:15:5
   |
15 | /// https://github.com/rust-lang/rust/issues/101103 for times when this
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use an automatic link instead: `<https://github.com/rust-lang/rust/issues/101103>`
   |
   = note: bare URLs are not automatically turned into clickable links
```

**Pull request merged**